### PR TITLE
backtester req

### PR DESCRIPTION
**Pull request by GuillaumeDSM**
requires https://github.com/Drakkar-Software/OctoBot-Commons/pull/303
requires https://github.com/Drakkar-Software/OctoBot-Tentacles-Manager/pull/148

### CODE REVIEW

File: octobot_backtesting/channels_manager.py [hunk 1]
```diff
@@ -67,20 +69,20 @@ def clear_empty_channels_producers(self):
             if producer.channel.get_consumers()
         ]
 
-    def clear_priority_levels(self):
-        self.priority_levels = [
-            priority_level.value
+    def clear_producers_by_priority_levels(self):
```
Comment (Herklos) [inline octobot_backtesting/channels_manager.py:72]: Shouldn't it be named `update_producers...`?

File: octobot_backtesting/channels_manager.py [hunk 2]
```diff
@@ -109,3 +143,23 @@ def _check_producers_consumers_emptiness(producers, priority_level):
         if not producer.is_consumers_queue_empty(priority_level):
             return False
     return True
+
+
+def _check_producers_has_priority_consumers(producers, priority_level):
+    for producer in producers:
+        for consumer in producer.channel.get_consumers():
+            if consumer.priority_level == priority_level:
+                return True
```
Comment (Herklos) [inline octobot_backtesting/channels_manager.py:152]: Should we use the new `get_prioritized_consumers` and check if its len == 0 ?

File: octobot_backtesting/channels_manager.py [hunk 3]
```diff
@@ -109,3 +143,23 @@ def _check_producers_consumers_emptiness(producers, priority_level):
...
+
+
+def _get_producers_with_priority_level_consumers(producers, priority_level):
+    return [
+        producer
+        for producer in producers
+        if any(
+            consumer
+            for consumer in producer.channel.get_consumers()
+            if consumer.priority_level <= priority_level
```
Comment (Herklos) [inline octobot_backtesting/channels_manager.py:163]: Same here with the new channel. get_prioritized_consumers()

File: octobot_backtesting/backtest_data.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+                exchange_name=exchange,
+                symbol=symbol,
+                time_frame=time_frame,
+                inferior_timestamp=start_timestamp,
+            )
+        return [
+            candle[-1]
+            for candle in all_candles
+            if candle[-1][commons_enums.PriceIndexes.IND_PRICE_TIME.value] <= end_timestamp
+        ]
```
Comment (Herklos) [inline octobot_backtesting/backtest_data.py:50]: Should we add a superior_timestamp instead of processing the returned data?

Comment (GuillaumeDSM) [reply octobot_backtesting/backtest_data.py:50]: "manually filter max candles not to change chronological cache behavior" in comments above :)

Comment (Herklos) [reply octobot_backtesting/backtest_data.py:50]: Ok that's right, my bad 😄

**Review by Herklos — APPROVED** [review]
👍

**Pull request merged**